### ISSUE TITLE
BUG: unable to securely deploy app as scipy 1.14.0 requires write permission in docker container.

### ISSUE THREAD

**Issue by Cemlyn** [scipy/scipy#21079]
### Describe your issue.

I am unable to securely deploy an app using scipy 1.14.0.

For security reasons, docker containers are run in read-only mode. Up to now I've been using scipy 1.11.4 and have experienced no issues. Upgrading to 1.14.0 causes my deployments to fail as scipy is trying to create a temporary folder but cannot.

This example repo shows how i am using scipy within my container: https://github.com/Cemlyn/scipy-example. Of course I build and run this simple app by running the following:
`docker build . -t test`
`docker run -it --read-only test`

Is it possible to revert to behaviour previously seen in version 1.11.4?

### Reproducing Code Example

```python
from scipy.stats import norm

means = [0,1,2,3]
stds = [1,1,2,2]

normal_dists = norm(means,stds)

print(normal_dists.mean)
```

### Error message

```shell
(temp2) ccr@Cemlyns-MBP scipy-bug % docker run -it --read-only test
Traceback (most recent call last):
  File "messagestream.pyx", line 91, in scipy._lib.messagestream.MessageStream.close
TypeError: remove: path should be string, bytes or os.PathLike, not NoneType
Exception ignored in: 'scipy._lib.messagestream.MessageStream.__dealloc__'
Traceback (most recent call last):
  File "messagestream.pyx", line 91, in scipy._lib.messagestream.MessageStream.close
TypeError: remove: path should be string, bytes or os.PathLike, not NoneType
Traceback (most recent call last):
  File "/opt/pysetup/test_script.py", line 1, in <module>
    from scipy.stats import norm
  File "/opt/pysetup/.venv/lib/python3.11/site-packages/scipy/stats/__init__.py", line 628, in <module>
    from ._rvs_sampling import rvs_ratio_uniforms
  File "/opt/pysetup/.venv/lib/python3.11/site-packages/scipy/stats/_rvs_sampling.py", line 2, in <module>
    from scipy.stats.sampling import RatioUniforms
  File "/opt/pysetup/.venv/lib/python3.11/site-packages/scipy/stats/sampling.py", line 59, in <module>
    from ._sampling import FastGeneratorInversion, RatioUniforms  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pysetup/.venv/lib/python3.11/site-packages/scipy/stats/_sampling.py", line 8, in <module>
    from ._unuran.unuran_wrapper import NumericalInversePolynomial
  File "unuran_wrapper.pyx", line 188, in init scipy.stats._unuran.unuran_wrapper
  File "unuran_wrapper.pyx", line 167, in scipy.stats._unuran.unuran_wrapper._setup_unuran
  File "messagestream.pyx", line 29, in scipy._lib.messagestream.MessageStream.__cinit__
  File "/usr/local/lib/python3.11/tempfile.py", line 351, in mkstemp
    prefix, suffix, dir, output_type = _sanitize_params(prefix, suffix, dir)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tempfile.py", line 129, in _sanitize_params
    dir = gettempdirb()
          ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tempfile.py", line 320, in gettempdirb
    return _os.fsencode(_gettempdir())
                        ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tempfile.py", line 309, in _gettempdir
    tempdir = _get_default_tempdir()
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tempfile.py", line 224, in _get_default_tempdir
    raise FileNotFoundError(_errno.ENOENT,
FileNotFoundError: [Errno 2] No usable temporary directory found in ['/tmp', '/var/tmp', '/usr/tmp', '/opt/pysetup']
```

### SciPy/NumPy/Python version and system information

```shell
{
  "Compilers": {
    "c": {
      "name": "clang",
      "linker": "ld64",
      "version": "15.0.0",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "3.0.10",
      "commands": "cython"
    },
    "c++": {
      "name": "clang",
      "linker": "ld64",
      "version": "15.0.0",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld64",
      "version": "13.3.0",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.16.1",
      "include directory": "../../../../../../private/var/folders/b3/2xm02wpd21qgrpkck5q1c6k40000gn/T/pip-build-env-q_06ff8x/overlay/lib/python3.11/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "aarch64",
      "family": "aarch64",
      "endian": "little",
      "system": "darwin"
    },
    "build": {
      "cpu": "aarch64",
      "family": "aarch64",
      "endian": "little",
      "system": "darwin"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "accelerate",
      "found": true,
      "version": "unknown",
      "detection method": "extraframeworks",
      "include directory": "unknown",
      "lib directory": "unknown",
      "openblas configuration": "unknown",
      "pc file directory": "unknown"
    },
    "lapack": {
      "name": "accelerate",
      "found": true,
      "version": "unknown",
      "detection method": "extraframeworks",
      "include directory": "unknown",
      "lib directory": "unknown",
      "openblas configuration": "unknown",
      "pc file directory": "unknown"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.12.0",
      "detection method": "config-tool",
      "include directory": "unknown"
    }
  },
  "Python Information": {
    "path": "/private/var/folders/b3/2xm02wpd21qgrpkck5q1c6k40000gn/T/cibw-run-ut6da4bl/cp311-macosx_arm64/build/venv/bin/python",
    "version": "3.11"
  }
}
```

**Comment by nickodell** [comment]
Here's what I've found from investigating this bug. I've not looked at this code before, so take this with a grain of salt.

1. SciPy stats loads a library called unuran for the purpose of generating random numbers.
1. It appears that unuran logs all of its errors to a `FILE *`.
2. That's not the interface we want to provide to users, so we capture that into either an `open_memstream()` (an in-memory `FILE *` object) or a tempfile, depending on what the operating system supports, so we can turn the error log stream into Exception objects. 

   This means that SciPy could accommodate your use case of running on a non-writable filesytem, as long as it can use `open_memstream()`.
3. `strace` doesn't seem to provide any smoking guns for why `open_memstream` couldn't be used. I'm using Linux to test this, so there's definitely operating system support.
4. In my dev build, the define `HAVE_OPEN_MEMSTREAM` is set to 0, despite being on a platform which supports `open_memstream()`. To double-check that I have this support, I tried the `open_memstream()` example program on [this page](https://www.gnu.org/software/libc/manual/html_node/String-Streams.html), and that programs works.
5. When I manually patch `meson.build` to set `HAVE_OPEN_MEMSTREAM` to 1, then it uses memstream, and this fixes the reproducer case.

   ```diff
   diff --git a/scipy/_lib/meson.build b/scipy/_lib/meson.build
   index fe1252bd3..1b0a7e473 100644
   --- a/scipy/_lib/meson.build
   +++ b/scipy/_lib/meson.build
   @@ -85,13 +85,7 @@ py3.extension_module('_test_deprecation_def',
    #       compiler? Does it not exist, or is it hidden elsewhere?
    
    conf_memstream = configuration_data()
   -
   -if meson.get_compiler('c').has_function('open_memstream',
   -                                        prefix: '#include <stdio.h>')
   -  conf_memstream.set('has_openmemstream', '1')
   -else
   -  conf_memstream.set('has_openmemstream', '0')
   -endif
   +conf_memstream.set('has_openmemstream', '1')
    
    configure_file(
      input: 'src/messagestream_config.h.in',
   ```

   It looks like something is wrong with this feature detection. But it seems fine according to the docs.

   I ran the tests, and all of them pass with `HAVE_OPEN_MEMSTREAM` force-set to 1. (Of course, we can't do this for all platforms. Windows, for example, does not support `open_memstream()`.)

7. To check my assumptions that glibc has this function, I tried writing a file called `test.c` with just the following:

   ```
   #include <stdio.h>
   ```

   I then ran `gcc -E test.c` to expand the `#include`. In the expanded output, there is a definition of `open_memstream`.

8. I also looked into whether this is something that only I am facing, or if this is also happening for published wheels. I wrote a script to download wheels from pypi from 1.11.0 through 1.14.0, (not including patch versions) and search it for `messagestream_open_memstream`, and disassemble that function. In every case, the code looks like this:

   ```
   000000000000c920 <messagestream_open_memstream>:
       c920:	31 c0                	xor    %eax,%eax
       c922:	c3                   	ret
   ```

   In other words, it returns NULL for this function. This indicates that in the following code, `HAVE_OPEN_MEMSTREAM` is set to 0 when building wheels.

   https://github.com/scipy/scipy/blob/ab7d08c6148286059f6498ab5c3070268d13cbd9/scipy/_lib/src/messagestream.h#L8-L18
9. In order to try to find why meson thinks that I don't have `open_memstream()`, I looked at the meson logs, and found what options it is passing to the compiler. Removing the options one at a time, I found that passing the `-std=c11` option causes glibc to hide the `open_memstream`, so meson thinks it is missing. See also [here](https://stackoverflow.com/questions/53242997/why-does-std-c11-with-gcc-hide-popen-in-stdio-h).
10. While searching through the project for uses of `_GNU_SOURCE`, I found that this has already been reported, but not fixed.  I really ought to have searched for previous reports sooner before spending eight hours debugging this, but you live and learn. https://github.com/scipy/scipy/issues/19911
12. I think I have a reasonably simple patch to fix this:
   
    ```diff
    diff --git a/scipy/_lib/meson.build b/scipy/_lib/meson.build
    index fe1252bd3..3ed2de0a6 100644
    --- a/scipy/_lib/meson.build
    +++ b/scipy/_lib/meson.build
    @@ -87,7 +87,7 @@ py3.extension_module('_test_deprecation_def',
    conf_memstream = configuration_data()

    if meson.get_compiler('c').has_function('open_memstream',
    -                                        prefix: '#include <stdio.h>')
    +                                        prefix: '#define _POSIX_C_SOURCE 200809L\n#include <stdio.h>')
      conf_memstream.set('has_openmemstream', '1')
    else
      conf_memstream.set('has_openmemstream', '0')
    ```

    It doesn't appear to be necessary to add the same define to `messagestream.h`. I'm guessing that it gets imported after a bunch of Python stuff, and that already sets `_POSIX_C_SOURCE` or something similar.

    This seems to be enough to fix your repro case.

**Comment by nickodell** [comment]
As previously mentioned, this is a duplicate of [gh-19911](https://github.com/scipy/scipy/issues/19911)

**Comment by nickodell** [comment]
Hi Cemlyn, I wrote a PR to fix this issue, and it's been accepted. It's planned to be included in the SciPy 1.15.0 update. It may also be backported to version 1.14.1, as it's marked as a backport candidate.

**Comment by Cemlyn** [comment]
Thanks for getting a fix in so quickly - much appreciated!
